uaclient/contract.py

```diff
@@ -183,32 +194,33 @@ def request_updated_contract(cfg, contract_token=None, allow_enable=False):
             'Got unexpected contract_token on an already attached machine')
     contract_client = UAContractClient(cfg)
     if contract_token:  # We are a mid ua-attach and need to get machinetoken
-        try:
-            new_token = contract_client.request_contract_machine_attach(
-                contract_token=contract_token)
-        except util.UrlError:
-            return False
+        new_token = contract_client.request_contract_machine_attach(
+            contract_token=contract_token)
     else:
         machine_token = orig_token['machineToken']
         contract_id = orig_token['machineTokenInfo']['contractInfo']['id']
+        new_token = contract_client.request_machine_token_refresh(
+            machine_token=machine_token, contract_id=contract_id)
+    user_errors = []
+    for name, entitlement in sorted(cfg.entitlements.items()):
+        if entitlement['entitlement'].get('entitled'):
+            # Obtain each entitlement's accessContext for this machine
+            new_access = contract_client.request_resource_machine_access(
+                new_token['machineToken'], name)
+        else:
+            new_access = entitlement
         try:
-            new_token = contract_client.request_machine_token_refresh(
-                machine_token=machine_token, contract_id=contract_id)
-        except util.UrlError:
-            return False
-    try:
-        for name, entitlement in sorted(cfg.entitlements.items()):
-            if entitlement['entitlement'].get('entitled'):
-                # Obtain each entitlement's accessContext for this machine
-                new_access = contract_client.request_resource_machine_access(
-                    new_token['machineToken'], name)
-            else:
-                new_access = entitlement
             process_entitlement_delta(
                 orig_entitlements.get(name, {}), new_access,
                 allow_enable=allow_enable)
-    except util.UrlError as e:
-        logging.error(
-            'Could not obtain updated contract information. %s', str(e))
-        return False
-    return True
+        except exceptions.UserFacingError as e:
+            user_errors.append(e)
+        except Exception as e:
+            with util.disable_log_to_console():
+                logging.exception(str(e))
+            raise exceptions.UserFacingError(
+                'Unexpected error handling Ubuntu Advantage contract changes')
+    if user_errors:
+        raise exceptions.UserFacingError(
+            'Failure processing Ubuntu Advantage contract changes. %s' %
```

Comment:
I think this will format a bit weirdly. With `user_errors = ['a', 'b']`, I think it would output:

```
Failure processing Ubuntu Advantage contract changes. a
b
```

which looks wonky to me. Perhaps:

```
Failure processing Ubuntu Advantage contract changes:
- a
- b
```

would be better?

Reply:
done.